src/panels/lovelace/components/hui-action-editor.ts

```diff
@@ -33,15 +41,18 @@ export class HuiActionEditor extends LitElement {
   }
 
   get _navigation_path(): string {
-    return this.config!.navigation_path || ""; // TODO Why is this only seeing ToggleActionConfig?
+    const config = this.config! as NavigateActionConfig;
+    return config.navigation_path || "";
   }
 
   get _service(): string {
-    return this.config!.service || ""; // TODO Why is this only seeing ToggleActionConfig?
+    const config = this.config! as CallServiceActionConfig;
+    return config.service || "";
   }
 
   get _service_data(): { [key: string]: any } {
-    return {}; // TODO Need advice on how to handle this free-form data
+    const config = this.config! as CallServiceActionConfig;
+    return config.service_data || {};
```

Comment:
This won't work. You need to `JSON.stringify` it for putting it in a textarea, then when user is done, it needs to be `JSON.parse` again. If it's invalid JSON, we should not be allowed to save.

Reply:
Yeah my first attempt in the commit before tried that but tried to apply it after every key stroke and quickly got out of whack. Will revisit in a future PR perhaps.